internal/characters/nahida/skill.go

```diff
@@ -0,0 +1,261 @@
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/core/glog"
+	"github.com/genshinsim/gcsim/pkg/enemy"
+)
+
+var skillPressFrames []int
+var skillHoldFrames []int
+
+func init() {
+	skillPressFrames = frames.InitAbilSlice(32)
+	skillPressFrames[action.ActionAttack] = 28
+	skillPressFrames[action.ActionCharge] = 28
+	skillPressFrames[action.ActionSkill] = 32
+	skillPressFrames[action.ActionBurst] = 32
+	skillPressFrames[action.ActionDash] = 27
+	skillPressFrames[action.ActionJump] = 26
+	skillPressFrames[action.ActionSwap] = 25
+
+	skillHoldFrames = frames.InitAbilSlice(63)
+	skillHoldFrames[action.ActionAttack] = 57
+	skillHoldFrames[action.ActionCharge] = 58
+	skillHoldFrames[action.ActionSkill] = 62
+	skillHoldFrames[action.ActionBurst] = 62
+	skillHoldFrames[action.ActionDash] = 59
+	skillHoldFrames[action.ActionJump] = 62
+	skillHoldFrames[action.ActionSwap] = 57
+}
+
+const (
+	skillPressCD        = 300
+	skillHoldCD         = 360
+	skillPressHitmark   = 13
+	skillMarkKey        = "nahida-tri-karma"
+	skillICDKey         = "nahida-tri-karma-icd"
+	triKarmaParticleICD = "nahida-tri-karma-particle-icd"
+)
+
+func (c *char) Skill(p map[string]int) action.ActionInfo {
+	if p["hold"] == 0 {
+		return c.skillPress(p)
+	} else {
+		return c.skillHold(p)
+	}
+}
+
+func (c *char) skillPress(p map[string]int) action.ActionInfo {
+
+	ai := combat.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "All Schemes to Know (Press)",
+		AttackTag:  combat.AttackTagElementalArt,
+		ICDTag:     combat.ICDTagNone,
+		ICDGroup:   combat.ICDGroupDefault,
+		StrikeType: combat.StrikeTypeDefault,
+		Element:    attributes.Dendro,
+		Durability: 25,
+		Mult:       skillPress[c.TalentLvlSkill()],
+	}
+
+	c.Core.QueueAttack(
+		ai,
+		combat.NewCircleHit(c.Core.Combat.Player(), 5),
+		0, //TODO: snapshot delay?
+		skillPressHitmark,
+		c.skillMarkTargets,
+	)
+
+	c.SetCDWithDelay(action.ActionSkill, skillPressCD, 11)
+
+	return action.ActionInfo{
+		Frames:          frames.NewAbilFunc(skillPressFrames),
+		AnimationLength: skillPressFrames[action.InvalidAction],
+		CanQueueAfter:   skillPressFrames[action.ActionSwap], // earliest cancel
+		State:           action.SkillState,
+	}
+}
+
+func (c *char) skillHold(p map[string]int) action.ActionInfo {
+	hold := p["hold"]
+	if hold > 330 {
+		hold = 330
+	}
+	if hold < 30 {
+		hold = 30 //TODO: hold appears to have a min before the camera appears
+	}
+
+	ai := combat.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "All Schemes to Know (Hold)",
+		AttackTag:  combat.AttackTagElementalArt,
+		ICDTag:     combat.ICDTagNone,
+		ICDGroup:   combat.ICDGroupDefault,
+		StrikeType: combat.StrikeTypeDefault,
+		Element:    attributes.Dendro,
+		Durability: 25,
+		Mult:       skillHold[c.TalentLvlSkill()],
+	}
+
+	c.Core.QueueAttack(
+		ai,
+		combat.NewCircleHit(c.Core.Combat.Player(), 5),
```

Comment:
should be radius 25 if you want to assume you get every enemy in the Hold E range